packages/logger/src/logger.ts

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This is the file that people using Node.js will actually import. You should
+ * only include this file if you have something specific about your
+ * implementation that mandates having a separate entrypoint. Otherwise you can
+ * just use index.ts
+ */
+
+/**
+ * A container for all of the Logger instances
+ */
+export const instances: Logger[] = [];
+
+/**
+ * The JS SDK supports 5 log levels and also allows a user the ability to
+ * silence the logs altogether.
+ *
+ * The order is a follows:
+ * DEBUG < VERBOSE < INFO < WARN < ERROR < SILENT
+ *
+ * All of the log types above the current log level will be captured (i.e. if
+ * I set the log level to `INFO`, errors will still be logged, but `DEBUG` and
+ * `VERBOSE` logs will not)
+ */
+export enum LogLevel {
+  DEBUG,
+  VERBOSE,
+  INFO,
+  WARN,
+  ERROR,
+  SILENT
+}
+
+/**
+ * A container for the default log level
+ */
+let defaultLogLevel: LogLevel = LogLevel.WARN;
+
+/**
+ * A function to set the default log level externally
+ */
+export function setDefaultLogLevel(val: LogLevel) {
+  if (!(val in LogLevel)) {
+    throw new TypeError('Attempted to Invalid value assigned to `logLevel`');
+  }
+  defaultLogLevel = val;
```

Comment:
Hrm...  So if you call setDefaultLogLevel(), which I assume will be the public API we eventually expose to users, shouldn't we notify all of the existing instances?  As a user I should be able to enable / disable logging at will.

I'm fuzzy on how the global and per-component log levels are meant to interact. E.g. what happens if I set the default one, then set a per-component one, then set the default one again?

Honestly I would just simplify and only have the default / global log level for now.  But if you keep both, please make sure they interact in a sane way (taking into consideration how you expect to expose the API to the end user eventually).

Reply:
I would go even further and suggest that we only need to set the default log level for the Firebase logger. If a user specifies their own logger, they can strip out log messages as they deem necessary.

Reply:
I've tried to make it so that each log client can set their own log levels (e.g. Database and Firestore, currently, can't rely on a single log level as they are handled separately). Though there are other ways you could solve this, I've opted to just keep the LogLevel, local to the Logger instance.

I've removed the global default thing, so that is gone.

There is still a top level "set-all" type of functionality. This will allow us to do global level sets across all `Logger` instances which we will need going forward.